core/descriptors/src/org/jetbrains/kotlin/descriptors/impl/FunctionDescriptorImpl.java

```diff
@@ -606,6 +625,27 @@ protected FunctionDescriptor doSubstitute(@NotNull CopyConfiguration configurati
         );
         if (substitutor == null) return null;
 
+        List<ReceiverParameterDescriptor> substitutedContextReceiverParameters = new ArrayList<ReceiverParameterDescriptor>();
+        if (!configuration.newContextReceiverParameters.isEmpty()) {
+            for (ReceiverParameterDescriptor newContextReceiverParameter : configuration.newContextReceiverParameters) {
+                KotlinType substitutedContextReceiverType =
+                        substitutor.substitute(newContextReceiverParameter.getType(), Variance.IN_VARIANCE);
+                if (substitutedContextReceiverType == null) {
+                    return null;
+                }
+                ReceiverParameterDescriptor substitutedContextReceiverParameter = new ReceiverParameterDescriptorImpl(
+                        substitutedDescriptor,
+                        new ExtensionReceiver(
+                                substitutedDescriptor, substitutedContextReceiverType, newContextReceiverParameter.getValue()
+                        ),
+                        newContextReceiverParameter.getAnnotations()
+                );
```

Comment:
We should create `ContextReceiver`, not `ExtensionReceiver`. + there is a specific function which can do this